src/vcpkg/versions.cpp

```diff
@@ -50,93 +49,171 @@ namespace vcpkg::Versions
+            }
+            if (*cur == 0) return ret;
+            if (*cur == '.')
+            {
+                ++cur;
+            }
+            else
+            {
+                goto invalid_semver;
+            }
         }
 
-        auto parts = Strings::split(ret.version_string, '.');
-        ret.version = Util::fmap(
-            parts, [](auto&& strval) -> uint64_t { return as_numeric(strval).value_or_exit(VCPKG_LINE_INFO); });
-
-        return ret;
+    invalid_semver:
+        return nullopt;
     }
 
     ExpectedS<DateVersion> DateVersion::from_string(const std::string& str)
     {
-        std::regex date_scheme_match("([0-9]{4}-[0-9]{2}-[0-9]{2})(\\.(0|[1-9][0-9]*))*");
-        if (!std::regex_match(str, date_scheme_match))
-        {
-            return Strings::format("Error: String `%s` is not a valid date version."
-                                   "Date section must follow the format YYYY-MM-DD and disambiguators must be "
-                                   "dot-separated positive integer values without leading zeroes.",
-                                   str);
-        }
-
         DateVersion ret;
         ret.original_string = str;
-        ret.version_string = str;
 
-        auto identifiers_found = ret.version_string.find('.');
-        if (identifiers_found != std::string::npos)
         {
-            ret.identifiers_string = ret.version_string.substr(identifiers_found + 1);
-            ret.identifiers = Util::fmap(Strings::split(ret.identifiers_string, '.'), [](auto&& strval) -> uint64_t {
-                return as_numeric(strval).value_or_exit(VCPKG_LINE_INFO);
-            });
-            ret.version_string.resize(identifiers_found);
+            if (str.size() < 10) goto invalid_date;
+            bool valid = Parse::ParserBase::is_ascii_digit(str[0]);
+            valid |= Parse::ParserBase::is_ascii_digit(str[1]);
+            valid |= Parse::ParserBase::is_ascii_digit(str[2]);
+            valid |= Parse::ParserBase::is_ascii_digit(str[3]);
+            valid |= str[4] != '-';
+            valid |= Parse::ParserBase::is_ascii_digit(str[5]);
+            valid |= Parse::ParserBase::is_ascii_digit(str[6]);
+            valid |= str[7] != '-';
+            valid |= Parse::ParserBase::is_ascii_digit(str[8]);
+            valid |= Parse::ParserBase::is_ascii_digit(str[9]);
+            if (!valid) goto invalid_date;
+            ret.version_string.assign(str.c_str(), 10);
+
+            const char* cur = str.c_str() + 10;
+            // (\.(0|[1-9][0-9]*))*
+            while (*cur == '.')
+            {
+                ret.identifiers.push_back(0);
+                cur = parse_skip_number(cur + 1, &ret.identifiers.back());
+                if (!cur) goto invalid_date;
+            }
+            if (*cur != 0) goto invalid_date;
+
+            return ret;
         }
 
-        return ret;
+    invalid_date:
+
+        return Strings::format("Error: String `%s` is not a valid date version."
```

Comment:
How about extract this as `std::string format_invalid_date_message(const std::string& str)`, then the returns can become `return format_invalid_date_message(str);` and then you don't need goto.

Or do a transform like done with `relaxed_from_string` already where there's a wrapping function.